Added octrees and kd-trees as partitioners

saada2006 opened the pull request:
# Description

- Changed `UniversePartitioner` to be a virtual object
- Turned the old z-plane partitioner object into a subclass of `UniversePartitioner` called `ZPlanePartitioner`
- Added octree partitioner in class `OctreePartitioner` that is subclass of `UniversePartitioner`. It has fast construction and traversal, as well as exporting to/importing from files
     - Added file format for exporting/importing partitioners, see `/examples/atr/omcp_file_format.md` 
     - Octree uses z-plane partitioner as fallback when a cell is not found by `Universe::find_cell`
- Added kd tree partitioner in class `KdTreePartitioner` that is subclass of `UniversePartitioner`. Construction and traversal times are slightly worse than octrees, and currently it does not support importing to/exporting from files.
     - Like the octree, the kd tree uses z-plane partitioner as fallback when a cell is not found by `Universe::find_cell`
- Added Advanced Test Reactor as example (see `examples/atr/`)

From my testing on my i9-11900k using 16 threads and the ATR example with 30 batches (10 inactive) and 40000 particles, the time spent in `Universe::find_cell` went from 38% with the `ZPlanePartitioner` to 0.75% on the `OctreePartitioner` and 1.2% on the `KdTreePartitioner`. Octree construction should take approximately 13 seconds, and kd tree construction should take 15 seconds. 

I removed performance statistic on how much time is spent in `Universe::find_cell` on the last few commits for a cleaner output. You can get them back via cloning the repo and moving back to commit `231bbae85a8b59be759ad1f16634dec23caca35f`. 

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)
- [ ] ~~I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)~~
- [ ] ~~I have made corresponding changes to the documentation (if applicable)~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works (if applicable)~~

gridley reviewed (changes requested):
Looks good Saad!

One thing we'll need for sure to get this merged is some capability for users to manually set the partitioner that they'd like to be used for a given universe in the geometry.xml file. In fact, it will likely be best to have _no_ partitioner used by default, in most problems. It is quite common to have small universes with only one or two cells, for example in problems that have tiled geometry.

Another likely necessity (although this depends on the opinion of other reviewers) will be not using a binary file for the partitioner. It should be straightforward to use the tools in hdf_interface to write it to HDF5.

Many of the classes you've introduced here also have no comments on what they're used for. In particular, it would be nice to have some brief explainers on how the compressed nodes work in their headers.

The final remark is about the pool allocator. If that is being used by multiple threads, how do we know that it's working correctly and not redundantly allocating or leaking? This seems like a hazard for anyone to touch later on. Generally we are not doing manual memory management anywhere else. It might be best to use unique_ptrs there so as to remove the need to call delete[] in the end, and just return NodeT* from unique_ptr::get as an observing pointer. Similarly I'd recommend the use of make_unique for the individual allocations after making that change.

saada2006 commented on CMakeLists.txt:
```diff
@@ -462,6 +468,7 @@ target_include_directories(libopenmc
 )
 
 # Set compile flags
+list(APPEND cxxflags "-O3")
```
For some weird reason, it doesn't set `-O3` when I compile with `-DCMAKE_BUILD_TYPE=Release` on my end. This is the script I use to automatically build OpenMC and run the ATR example. 
```bash
cd build
cmake -DCMAKE_BUILD_TYPE=Release .. 
make -j 16
sudo make install # update our old build with our new build

echo "Running ATR test..."
cd ../examples/atr/
python3 simulation.py
```

gridley replied:
That is strange! What system are you on? Also, can you still remove this line, then?

saada2006 commented on include/openmc/kdtree_partitioner.h:
```diff
@@ -0,0 +1,63 @@
+#include "partitioner_utils.h"
+#include "universe.h"
+#include "zplane_partitioner.h"
+#include <vector>
```
git-clang-format sometimes gives me improper results. It might be better for someone to clone the repo after I've fixed everything else to run clang-format on their end.

saada2006 commented on include/openmc/partitioner_utils.h:
```diff
@@ -0,0 +1,322 @@
+  void extend(const AABB& other_box);
+
+  vec3 get_center() const;
+  float surface_area() const;
```
I switched to double precision in most areas. In some areas, I kept single precision for better cache efficiency. 

I think it would be nice to turn `Position` into a template, where `Postion<double>` means you want double precision (the default if no type is supplied), `Position<float>` means you want single precision, etc. It would help in certain areas (e.g. the kd tree partitioner) where double precision is not required and memory/cache efficiency is a bigger concern.

saada2006 commented on include/openmc/partitioner_utils.h:
```diff
@@ -0,0 +1,322 @@
+  }
+
+private:
+  std::vector<NodeT*> pools;
```
Made the node allocator thread safe and switched over to smart pointers.

saada2006 commented on include/openmc/partitioner_utils.h:
```diff
@@ -0,0 +1,322 @@
+  bool operator<(const CellPoint& other) const;
+};
+
+class Bin {
```
Added comments in `kdtree_partitioner.h/cpp`, `octree_partitioner.h/cpp`, and `partitioner_utils.h/cpp`.

saada2006 commented on include/openmc/partitioner_utils.h:
```diff
@@ -0,0 +1,322 @@
+
+  float compute_score()
+  {
+    const float REFINEMENT_BIN_SCORE_FOUND_MULT = 1.0;
```
Switched over to `constexpr` from `const` in most areas.

saada2006 commented on include/openmc/zplane_partitioner.h:
```diff
@@ -0,0 +1,37 @@
+
+using namespace std;
+
+class ZPlanePartitioner : public UniversePartitioner {
```
Does the ATR example have a good amount of x- and y-planes? If not, could you send an example input that does? I think it would be useful when testing out the efficiency of x- and y-plane partitioners.

saada2006 commented:
For the ATR example, Universe::bounding_box was returning an example that spanned from (-infinity, -infinity, -infinity) to (infinity, infinity, infinity). I cannot create an octree/kd tree with that as a bounding box, however I found setting the bounding box to be from (-130, -130, -130) to (130, 130, 130), very few calls to Universe::find_cell were out of bounds. I was thinking that a possible way to determine a good bounding box (if not manually set by the user) would be to extend the universe's bounding box when generating it on a particular axis by a value only if that value is not equal to infinity. Would that be a good idea?

saada2006 commented:
I've added bin grid partitioners and the partitioner bounds are no longer hardcoded by rather determined by the `Universe` class. Bin grid partitioners construct really fast (~3 seconds) and offer speeds exactly the same as octrees. I also spent time speeding up kd tree construction so that they are faster than octrees.

saada2006 commented:
You can now set partitioners from geometry.xml. Here is an example:
```xml
  <partitioner type="octree" bounds="-130 130 -130 130 -130 130" target-cells="6" universe="0" />
```
The values for `type` are `zplane`, `octree`, `kdtree`, `bin-grid`, and `auto`. `auto` defaults to the bin grid partitioner. 

`bounds` specifies the bounds of the octree in `xmin xmax ymin ymax zmin zmax` format. If the value of `bounds` is `auto` or there is no value for `bounds` set, the universe automatically generates a bounding box for the partitioner that ignores infinitely long distances.

If `octree` is selected for `type`, then an additional `target-cells` parameter will be available, which sets the target cells per node parameter in the octree constructor. If no value is supplied, then the parameter will default to 6. 

If `kdtree` is selected for `type`, then an additional `max-depth` parameter will be available, which sets the max depth parameter in the kd tree constructor. If no value is supplied, then the parameter will default to 16. 

If `bin-grid` is selected for `type`, then an additional `grid-res` parameter will be available, which sets the grid res parameter in the bin grid constructor. If no value is supplied, then the parameter will default to 32.

All partitioners have a `universe` attribute which describes while universe they are associated with.

saada2006 commented:
How to import from HDF5:
```xml
  <partitioner path="path/to/partitioner.h5" />
```
How to export to HDF5:
```xml
  <partitioner type="..." bounds="..." other-parameters="..." export-path="path/to/export/to/after/construction" universe="0"/>
```

pshriwise commented:
Just stopping in to say this is really cool and I hope to look it over in the nearish future. 😎 🎊

saada2006 commented:
Sorry for the super late update, but are there any other requested changes (besides cleaning up the repository)?

gridley commented on .gitignore:
```diff
@@ -106,3 +106,9 @@ CMakeSettings.json
 *.pkl
+
+# Large fules from partitioner builds
+*.omcp
```
aren't these now hdf5 files? So this shouldn't be needed, right?

saada2006 replied:
Removed.

gridley commented on .gitignore:
```diff
@@ -106,3 +106,9 @@ CMakeSettings.json
+# Large fules from partitioner builds
+*.omcp
+# Building scripts in root folder
+debug-*.sh
```
I think only you use this, so might want to remove it.

saada2006 replied:
Removed.

gridley commented on CMakeLists.txt:
```diff
@@ -331,7 +332,9 @@ endif()
 #===============================================================================
 
 list(APPEND libopenmc_SOURCES
+  src/aabb.cpp
```
Let's expand the full word in the file name so it's not ambiguous.

gridley commented on examples/pincell/plot.py:
Can you remove this file? It's already well-documented how to do this in other places.

saada2006 replied:
Done.

gridley commented on examples/pincell/sim.py:
Again, can you remove this one? If we should add it, it would be better to add to the examples repo. There is already a lot of new content in this PR.

saada2006 replied:
Done.

gridley commented on include/openmc/aabb.h:
I think the best path to merging this would be modifying/extending the BoundingBox class to work with the partitioners, and possibly change its implementation a little. Sorry this is a pain, but it would be quite nice to not have what is essentially redundant code here.

The BoundingBox class should probably have just been implemented using Position rather than xmin, xmax etc in the first place.

saada2006 replied:
Done.

gridley commented on include/openmc/kdtree_partitioner.h:
```diff
@@ -0,0 +1,118 @@
+
+namespace openmc {
+
+// This ndoe contains all the information you would want to know about a node
```
node

saada2006 replied:
Fixed.

gridley reviewed (commented):
A few minor comments here, but yeah, we're mainly just waiting on other people to test this out and look over the code. It might be worth putting the source files into their own directory, too, but let's decide that when someone else comes to look this over.

saada2006 commented:
Alright, @gridley I implemented the changes you requested. Did I miss anything? 

I also updated and moved the partitioner file format documentation to the io_formats folder.

paulromano commented:
Thanks for this PR @saada2006! I tried this out on one of my huge fusion models and unfortunately it segfaulted for me. I'll have to dig into it more but here's the traceback for starters:
```
Thread 9 "openmc" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x155524fea640 (LWP 873881)]
openmc::Universe::find_cell_for_point (this=<optimized out>, cells_to_search=std::vector of length 1, capacity 512 = {...}, r=...) at /usr/include/c++/11/bits/unique_ptr.h:173
173	      pointer    _M_ptr() const { return std::get<0>(_M_t); }
(gdb) bt
#0  openmc::Universe::find_cell_for_point (this=<optimized out>, cells_to_search=std::vector of length 1, capacity 512 = {...}, r=...) at /usr/include/c++/11/bits/unique_ptr.h:173
#1  0x00001555552c4f72 in _ZN6openmc19binned_point_searchINS_21CellPointUncompressedEEESt6vectorIT_SaIS3_EERKNS_8UniverseERKNS_19UniversePartitionerERKNS_11BoundingBoxE._omp_fn.0(void) () at /home/romano/openmc/include/openmc/partitioner_utils.h:343
#2  0x00001555550dfc0e in ?? () from /lib/x86_64-linux-gnu/libgomp.so.1
#3  0x0000155554a94b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#4  0x0000155554b26a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

saada2006 commented:
How odd. Based on the length/capacity of the vector, there shouldn't be a problem with the bin vector. However, based on the fact the error comes from somewhere in `unique_ptr`, I suspect that maybe the the compiler changes the type of `univ` from `const Universe&` to `unique_ptr<Universe>` and that unique pointer gets corrupted somehow. Try printing the address of `univ` right before line 343 in `partitioner_utils.h` and see if it changes right before the seg fault.

paulromano commented:
@saada2006 The type of `univ` is deterministic, so that shouldn't be the issue. The use of `unique_ptr` is really from within the `find_cell_for_point` method, where you have `model::cells[i_cell]->contains(...)`. It looks like it's going out of bounds on that vector. When I change it to `model::cells.at(i_cell)`, I get the following:
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 18446744073709551615) >= this->size() (which is 328472)

Thread 3 "openmc" received signal SIGABRT, Aborted.
[Switching to Thread 0x155525bf0640 (LWP 88548)]
__pthread_kill_implementation (no_tid=0, signo=6, threadid=23455449679424) at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=23455449679424) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=23455449679424) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=23455449679424, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x0000155554a42476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x0000155554a287f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x0000155554ea2b9e in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x0000155554eae20c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x0000155554ead1e9 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#8  0x0000155554ead959 in __gxx_personality_v0 () from /lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00001555554e8884 in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#10 0x00001555554e8f41 in _Unwind_RaiseException () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#11 0x0000155554eae4cb in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x0000155554ea54cd in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#13 0x0000155555406d6c in std::vector<std::unique_ptr<openmc::Cell, std::default_delete<openmc::Cell> >, std::allocator<std::unique_ptr<openmc::Cell, std::default_delete<openmc::Cell> > > >::_M_range_check (__n=18446744073709551615, this=<optimized out>) at /usr/include/c++/11/bits/stl_vector.h:1073
#14 std::vector<std::unique_ptr<openmc::Cell, std::default_delete<openmc::Cell> >, std::allocator<std::unique_ptr<openmc::Cell, std::default_delete<openmc::Cell> > > >::at (__n=18446744073709551615, this=<optimized out>) at /usr/include/c++/11/bits/stl_vector.h:1094
#15 openmc::Universe::find_cell_for_point (this=<optimized out>, cells_to_search=std::vector of length 1, capacity 512 = {...}, r=...) at /home/romano/openmc/src/universe.cpp:87
#16 0x00001555552c4f72 in _ZN6openmc19binned_point_searchINS_21CellPointUncompressedEEESt6vectorIT_SaIS3_EERKNS_8UniverseERKNS_19UniversePartitionerERKNS_11BoundingBoxE._omp_fn.0(void) () at /home/romano/openmc/include/openmc/partitioner_utils.h:345
#17 0x00001555550dfc0e in ?? () from /lib/x86_64-linux-gnu/libgomp.so.1
#18 0x0000155554a94b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#19 0x0000155554b26a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

saada2006 commented:
Ah, that makes a lot more sense. It comes from this section of code:
```cpp
        // if we didn't find the point, we need to sample and reinsert
        if (point.cell_ == -1) {
          const auto& possible_cells =
            fallback.get_cells(point.pos_, dummy_dir);
          sample_bin->copy_untested_cells(possible_cells, untested_cells);
          point.cell_ = univ.find_cell_for_point(untested_cells, point.pos_);

          sample_bin->insert(point.cell_);
        } // else don't bother inserting
```
If we didn't find the cell in the bin, we resort to using the the list of cells we get from the fallback (usually a z-plane partitioner) to find the cell. 

For some reason, in this rare case we don't end up finding the cell from the fallback's list of cells, which causes `Universe::find_cell_for_point` to return -1. Since the program doesn't check that the returned cell ID is -1, it stores it in the bin's list of cells. The next time we select the bin, we first iterate over the bin's list of cell to determine which cell the point is located in. Since we have -1 stored in that list, it results in a segfault. 

To fix this, we will need to check the cell ID returned from the fallback's list of cells. If it is -1, we resort to a complete search. I'll implement this fix as soon as I get home.

saada2006 commented:
Bug should be fixed now.

gridley commented:
Thanks for the quick fix Saad! @paulromano are you able to share any of these models so I can help debug and get this beast merged?

As another point... I think we are approaching the day where it may make sense to put OpenMC source files into their own directories. For instance we might want `geometry`, `physics`, and `partitioners`, similar to how we already group tallies into `tallies`.
